lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -247,6 +248,176 @@ std::unique_ptr<Pass> createLowerToHWPass() {
+                       /*direction*/ port.direction,
+                       /*type*/ port.type,
+                       /*argNum*/ ports.size()});
+      if (port.direction == hw::PortDirection::OUTPUT)
+        partInstOutputs.push_back(inst->getResult(port.argNum));
+      else
+        partInstInputs.push_back(inst->getOperand(port.argNum));
+    }
+  };
+  // Handle all other operators.
+  auto addOther = [&](Operation *op) { assert(false && "Unimplemented"); };
+
+  // Aggregate the args/results into partition module ports.
+  for (Operation *op : toMove) {
+    if (auto inst = dyn_cast<InstanceOp>(op)) {
+      Operation *modOp = topLevelSyms.getDefinition(inst.moduleNameAttr());
+      assert(modOp && "Module instantiated should exist. Verifier should have "
+                      "caught this.");
+
+      if (auto mod = dyn_cast<MSFTModuleOp>(modOp)) {
+        addModuleLike(inst, mod.getPorts());
+      } else if (auto mod = dyn_cast<MSFTModuleExternOp>(modOp)) {
+        addModuleLike(inst, mod.getPorts());
+      }
+    } else {
+      addOther(op);
+    }
+  }
+
+  //*************
+  //   Construct the partition module and replace the design partition op.
+
+  // Build the module.
+  auto partMod =
+      OpBuilder::atBlockEnd(getOperation().getBody())
+          .create<hw::HWModuleOp>(loc, partOp.verilogNameAttr(), ports);
+  Block *partBlock = partMod.getBodyBlock();
+  partBlock->clear();
+  auto partBuilder = OpBuilder::atBlockEnd(partBlock);
+
+  // Replace partOp with an instantion of the partition.
+  auto partInst = OpBuilder(partOp).create<hw::InstanceOp>(
+      loc, partMod, partOp.getNameAttr(), partInstInputs, ArrayAttr(),
+      SymbolTable::getSymbolName(partOp));
+
+  // Replace original ops' outputs with partition outputs.
+  assert(partInstOutputs.size() == partInst.getNumResults());
+  for (size_t resNum = 0, e = partInstOutputs.size(); resNum < e; ++resNum)
+    partInstOutputs[resNum].replaceAllUsesWith(partInst.getResult(resNum));
+
+  //*************
+  //   Move the operations!
+
+  // Map the original operation's inputs to block arguments.
+  mlir::BlockAndValueMapping mapping;
+  assert(partInstInputs.size() == partBlock->getNumArguments());
+  for (size_t argNum = 0, e = partInstInputs.size(); argNum < e; ++argNum) {
+    mapping.map(partInstInputs[argNum], partBlock->getArgument(argNum));
+  }
+
+  // Since the same value can map to multiple outputs, compute the 1-N mapping
+  // here.
+  DenseMap<Value, SmallVector<int, 1>> resultOutputConnections;
+  for (size_t outNum = 0, e = partInstOutputs.size(); outNum < e; ++outNum)
+    resultOutputConnections[partInstOutputs[outNum]].push_back(outNum);
+
+  // Hold the block outputs.
+  SmallVector<Value, 64> outputs(partInstOutputs.size(), Value());
+
+  // Clone the ops into the partition block. Map the results into the module
+  // outputs.
+  for (Operation *op : toMove) {
+    auto *newOp = partBuilder.insert(op->clone(mapping));
```

Comment:
`op->clone` isn't ideal -- a move method would be better. The documentation for the `move` method, however, specifies that it only works within a function (which I assume is a region...?)

Reply:
> it only works within a function (which I assume is a region...?)

My reading of the implementation is it should work if the block is within the same region. "function" in the comment is probably a holdover from the early days of MLIR that were very focused on software functions with a graph of basic blocks, but I think any region holding op would work.

Reply:
So in this case, I don't know if the `move` family of methods will work...

Reply:
> > it only works within a function (which I assume is a region...?)
> 
> My reading of the implementation is it should work if the block is within the same region. "function" in the comment is probably a holdover from the early days of MLIR that were very focused on software functions with a graph of basic blocks, but I think any region holding op would work.

Yeah, that was my assumption as well.